Updated QuantumGraph to Qiskit 2.0

owaisishtiaqsiddiqui opened the pull request:
- Removed dependency of Qiskit Ignis.
- Quantum Graph is now compatible with Qiskit 2.0.
- Updated imports and backend usage to align with `Qiskit 2.x.x` and `qiskit-aer 0.17.x`.
- Replaced deprecated tomography and decomposition methods with current versions `(StateTomography, TwoQubitBasisDecomposer).`
- Switched from the deprecated u3 gate to the universal u gate.
- Included detailed debug statements and adjusted execution methods for clarity and compatibility.

owaisishtiaqsiddiqui committed:
Updated Quantum-Graph to Qiskit 2.0

Removed Qiskit-Ignis dependency and update the QuantumGraph to Qiskit 2.0.

A running tutorial file.
All test files are running.

owaisishtiaqsiddiqui committed:
Create readme.md

owaisishtiaqsiddiqui committed:
Update README.md

owaisishtiaqsiddiqui committed:
Update README.md

owaisishtiaqsiddiqui committed:
Update README.md

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/qiskit-community/QuantumGraph?pullRequest=12) <br/>All committers have signed the CLA.

owaisishtiaqsiddiqui committed:
Update README.md

quantumjim commented:
Hi @owaisishtiaqsiddiqui. You have made impressive progress. However, I do see an issue. Consider the following code

```
from quantumgraph import QuantumGraph

n = 4
qubits = QuantumGraph(n)

qubits.qc.h(0)

qubits.update_tomography()

for j in range(n):
    print('Single qubit variables for qubit ' + str(j) + ':')
    print(qubits.get_bloch(j))
    print()
```

For me this yields the output

```
Single qubit variables for qubit 0:
{'X': 0, 'Y': 0, 'Z': 1}

Single qubit variables for qubit 1:
{'X': 0, 'Y': 0, 'Z': 1}

Single qubit variables for qubit 2:
{'X': 0, 'Y': 0, 'Z': 1}

Single qubit variables for qubit 3:
{'X': 1, 'Y': 0, 'Z': 0}
```

So here it is looks like qubit 3 that has experienced the Hadamard, rather than qubit 1. I assume this comes from your section starting [here](https://github.com/owaisishtiaqsiddiqui/QuantumGraph/blob/27fc570ace78da2248d7defe3fad5f41a43bfa0c/quantumgraph/QuantumGraph.py#L216), and you have effectively reversed the qubit order.

owaisishtiaqsiddiqui commented:
I am using Qiskit 2.0 myself though. I will check and run the file again. Thanks for the update.

owaisishtiaqsiddiqui committed:
3 files updated

Tutorial.ipynb (tested and updated)
Test_QuantumGraph.ipynb (updated)
QuantumGraph.py (updated)

owaisishtiaqsiddiqui commented:
> Hi @owaisishtiaqsiddiqui. You have made impressive progress. However, I do see an issue. Consider the following code
> 
> ```
> from quantumgraph import QuantumGraph
> 
> n = 4
> qubits = QuantumGraph(n)
> 
> qubits.qc.h(0)
> 
> qubits.update_tomography()
> 
> for j in range(n):
>     print('Single qubit variables for qubit ' + str(j) + ':')
>     print(qubits.get_bloch(j))
>     print()
> ```
> 
> For me this yields the output
> 
> ```
> Single qubit variables for qubit 0:
> {'X': 0, 'Y': 0, 'Z': 1}
> 
> Single qubit variables for qubit 1:
> {'X': 0, 'Y': 0, 'Z': 1}
> 
> Single qubit variables for qubit 2:
> {'X': 0, 'Y': 0, 'Z': 1}
> 
> Single qubit variables for qubit 3:
> {'X': 1, 'Y': 0, 'Z': 0}
> ```
> 
> So here it is looks like qubit 3 that has experienced the Hadamard, rather than qubit 1. I assume this comes from your section starting [here](https://github.com/owaisishtiaqsiddiqui/QuantumGraph/blob/27fc570ace78da2248d7defe3fad5f41a43bfa0c/quantumgraph/QuantumGraph.py#L216), and you have effectively reversed the qubit order.

Dear @quantumjim I have updated the issues pointed out according to your request. Let me know if it satisfies your criteria. 😃

owaisishtiaqsiddiqui committed:
Updated Tutorial.ipynb

Corrected Qubits order

quantumjim commented:
Thanks! This has removed the issue for `set_bloch`. However, it seems that something similar is happening also for `set_relationship`.

Specifically, run

```
from quantumgraph import QuantumGraph

n = 5
graph = QuantumGraph(n)
relationships = {'XX':+1}
graph.set_relationship(relationships,0,1)
```

If I then run `graph.get_relationship(0,1)` I find that $\langle XX \rangle=0, \langle ZZ \rangle=1$ instead of the $\langle XX \rangle=1, \langle ZZ \rangle=0$ required. Instead I find the required result for `graph.get_relationship(3,4)`.

This suggests to me that there is another effective reversal of the qubit order when you do the two qubit tomography.

owaisishtiaqsiddiqui commented:
> Thanks! This has removed the issue for `set_bloch`. However, it seems that something similar is happening also for `set_relationship`.
> 
> Specifically, run
> 
> ```
> from quantumgraph import QuantumGraph
> 
> n = 5
> graph = QuantumGraph(n)
> relationships = {'XX':+1}
> graph.set_relationship(relationships,0,1)
> ```
> 
> If I then run `graph.get_relationship(0,1)` I find that ⟨ X X ⟩ = 0 , ⟨ Z Z ⟩ = 1 instead of the ⟨ X X ⟩ = 1 , ⟨ Z Z ⟩ = 0 required. Instead I find the required result for `graph.get_relationship(3,4)`.
> 
> This suggests to me that there is another effective reversal of the qubit order when you do the two qubit tomography.

Understood

owaisishtiaqsiddiqui committed:
Updated QuantumGraph.py

owaisishtiaqsiddiqui committed:
updated set_relationship in QuantumGraph.py  and Tutorial.ipynb

owaisishtiaqsiddiqui committed:
Added extra cell for qubit ordering in Test_QuantumGraph.ipynb

owaisishtiaqsiddiqui commented:
Dear @quantumjim I got this result. Please check the test_quantumgraph.ipynb

`{'XX': np.float64(0.9909552061255673), 'XY': np.float64(0.002338813006651442), 'XZ': np.float64(-0.00495712560524509), 
'YX': np.float64(0.003666444635346955), 'YY': np.float64(-0.0007234271456779962), 'YZ': np.float64(-0.00244939221803704), 'ZX': np.float64(-0.00289622451153272), 'ZY': np.float64(-0.0013714461717840819), 'ZZ': np.float64(0.0004981342763795416)}`

owaisishtiaqsiddiqui committed:
Update Test_QuantumGraph.ipynb

quantumjim commented:
Looks great! Thanks for your efforts. This closes #11

owaisishtiaqsiddiqui commented:
> Looks great! Thanks for your efforts. This closes #11

You are welcome. Thankyou for giving me a chance to contribute.
